pallets/pallet-auction/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+
+					// Calculate how many blocks the auction will over, if it shorter than specific
+					// length, then extend to specific length.
+					if T::ExtendedLength::get() >
+						auction_data.expiry_block -
+							frame_system::Pallet::<T>::current_block_number()
+					{
+						// Remove the old auction id
+						AuctionsExpiryBlock::<T>::mutate(auction_data.expiry_block, |auctions| {
+							auctions.retain(|id| *id != auction_id);
+						});
+
+						let new_expiry = frame_system::Pallet::<T>::current_block_number()
+							.saturating_add(T::ExtendedLength::get());
+						auction_data.expiry_block = new_expiry;
+
+						// Add the new block number to the storage.
+						AuctionsExpiryBlock::<T>::append(new_expiry, auction_id);
+					}
+
+					Self::deposit_event(Event::<T>::Bid { who: id, auction_id, price });
+
+					Ok(())
+				} else {
+					Err(Error::<T>::InvalidAuctionId.into())
+				}
+			})
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::cancel_auction())]
+		pub fn cancel_auction(origin: OriginFor<T>, auction_id: AuctionId) -> DispatchResult {
+			// Get the account id
+			let id = ensure_signed(origin)?;
+
+			// Ensure role is customer.
+			T::RoleManager::ensure_role(&id, Role::Customer)?;
+
+			// todo read the storage auctions
+			let auction_data =
+				Auctions::<T>::get(auction_id).ok_or(Error::<T>::InvalidAuctionId)?;
+
+			// Ensure the nft is belong to the correct owner.
+			T::NftManager::ensure_nft_is_valid(&id, auction_data.nft_id)?;
+
+			// Compare whether current price is upon the reserve price.
+			if let Some((bider, price)) = auction_data.current_bid {
+				if price >= auction_data.reserve.unwrap_or_default() {
+					return Err(Error::<T>::AuctionCancelFailed.into());
+				} else {
+					// Transfer back the money to the bider.
+					T::Bank::transfer(&T::BidsPoolAccount::get(), &bider, price)?;
+				}
+			}
+			// Changer the Nft state to free.
+			T::NftManager::change_nft_state(auction_data.nft_id, NftState::Free)?;
+			// Remove the auction from storage Auctions.
+			Auctions::<T>::remove(auction_id);
+
+			Self::deposit_event(Event::<T>::AuctionCanceled { auction_id });
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Get the auction id to store into the NextAuctionId.
+		pub(crate) fn next_auction_id() -> AuctionId {
+			NextAuctionId::<T>::mutate(|id| {
+				*id = id.wrapping_add(1);
+				*id
+			})
+		}
+
+		/// Get the auction done.
+		fn resolve_auction(auction_id: AuctionId, auction_data: AuctionData<T>) {
+			if let Some((bider, price)) = auction_data.current_bid {
+				let bid_pool = T::BidsPoolAccount::get();
+
+				if price >= auction_data.reserve.unwrap_or_default() {
+					// transfer money
+					let tax = T::AuctionSuccessFeePercentage::get() * price;
+					let rest = price.saturating_sub(tax);
+					let _ = T::Bank::transfer(&bid_pool, &auction_data.owner, rest);
+					if let Ok(treasury) = T::Bank::treasury() {
```

Comment:
Maybe we should have a "transfer_to_treasury" just so we don't need to do this everywhere